Create script to fix EoL repos on XO laptops (fixes #3)

jwflory opened the pull request:
This pull request attempts to fix issue #3.
# What it does

The script pulls down the GPG public keys used for Fedora 18 package signing. It imports those keys into the system. Afterwards, it pulls down the repo files that have the correct `baseurl`, deletes the old files, and replaces them with the new.

After doing this, the XO user should be able to use `yum` once again.
# Why it matters

This will allow students to interact with `yum` in the command line on the XO to install software. At the beginning of my HFOSS course, my XO laptop was a freshly flashed system. Because of this, the repo files in the system were made to point to the URLs used if Fedora 18 were still a supported release. However, that is not the case. F18 hit EoL status many years ago. This script just adjusts the system to use the accurate GPG keys and repo `baseurl` so it will work once again.
# Note: Code review

A few things to note:
- This is my first "real" bash script I've written, so I'd like feedback on syntax or "good behavior" vs. "bad behavior"
- I no longer have an XO laptop to test this script on, making it hard to verify that it works

Having someone else review my basic script would be useful. This is probably relevant for @ritjoe in teaching future versions of the course. If @Nolski or @brendan-w wanted to proof my bash script, that would be nice too.

Cheers, and thanks.

Nolski commented:
Cc @ryansb probably has some good input on bash scripts. I am also a noob at bash scripts

ignatenkobrain commented on scripts/xo-eol-repos/f18/fix_repos.sh:
```diff
@@ -0,0 +1,68 @@
+
+# Environment variables
+KEYSIGN_DIR=/etc/pki/rpm-gpg
+WORK_DIR=$HOME/tmp
```
I would like to have quotes here

ryansb replied:
So there's a `$TMPDIR` variable, which would be preferred to adding a new top-level dir in `~`.

ignatenkobrain commented on scripts/xo-eol-repos/f18/fix_repos.sh:
```diff
@@ -0,0 +1,68 @@
+# Pull down the obsolete keys for Fedora 18 and import them
+#     https://getfedora.org/keys/obsolete.html
+#     https://getfedora.org/keys/faq/
+wget https://getfedora.org/static/DE7F38BD.txt
```
minimal installations contains only `curl`. I would use it with flags `-L -O`

ignatenkobrain commented on scripts/xo-eol-repos/f18/fix_repos.sh:
```diff
@@ -0,0 +1,68 @@
+	mkdir $HOME/tmp
+fi
+
+cd $HOME/tmp
```
we defined `$WORK_DIR` above for this (don't forget quotes)

ignatenkobrain commented on scripts/xo-eol-repos/f18/fix_repos.sh:
```diff
@@ -0,0 +1,68 @@
+if [ ! -d "$WORK_DIR" ]
+	then
+	echo "Working directory does not exist, making tmp/ in home directory..."
+	mkdir $HOME/tmp
```
we defined `$WORK_DIR` above for this (don't forget quotes)

jflory7 commented:
@ignatenkobrain, @ryansb: Awesome, thanks so much for this feedback!! It's really useful. I've optimized the script based on your feedback. Really appreciate the time.

@ritjoe said he was going to review it sometime today as well, so I'll leave this open for him to review before merging it into the repository.

ryansb commented on scripts/xo-eol-repos/f18/fix_repos.sh:
```diff
@@ -0,0 +1,67 @@
+#    License, v. 2.0. If a copy of the MPL was not distributed with this
+#    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+
```
Forgot to mention my favorite (and soon to be yours) bash options.

`set -euxo pipefail`

See http://redsymbol.net/articles/unofficial-bash-strict-mode/ for more info. I add `-x` to the options recommended in that article because `-x` prints out commands as they execute so you can see how a script works and what parts take a long time (instead of figuring it out by `echo`ing).

ritjoe commented:
So, I've looked at a couple of the XOs I have with me today. One already has git, so is not that instructive because it could have gotten it when the F18 repos were still fresh.

The other one, however, did not yet have git, but when I issued:

yum install git

it went out and found it. (I answered "n" so as not yet to install it).

(meta comment: darn it, why does github support ZIP but not .tgz?)

just a mo, let me repack this archive ...

...ok, there.  Please take a look at these repo files from this F18 system that seems still to be able to find package mirrors.  

[yum.repos.d.zip](https://github.com/FOSSRIT/toolbox/files/273670/yum.repos.d.zip)

Not that I understand this well, but my rough guesstimate is the service at mirrors.fedoraproject.org is able to take the GET parameters and return a working mirror even for versions no longer current

mirrorlist=https://mirrors.fedoraproject.org/metalink?repo=fedora-$releasever&arch=$basearch

set, for instance, releasever=18 and basearch=i386 and one gets a metalink file full of XML leading in turn to yet more XML files, which rabbit hole I haven't completely followed down to exhaustion.

So, I'm still not sure why there is differential behavior--jflory, did you try to do, eg, 

yum install git

on your newly flashed XO before starting to modify the .repo files? I'm wondering if something got fouled up on your XO pre-flash that required these modifications.

The XOs I've got I'd like to try to archive their contents in some way to provide a basis for comparison, so I'm not going to reflash them just yet. There are XOs on campus that I can lay hands on to reflash and then to see how if I can reproduce the problems downloading stuff from the repos on them.

A confounding factor might also be just straight up not having a good network connection, which I think came up in our f2f discussions.

jflory7 commented:
@ritjoe Looking at the repo files, I do think those are the same as the ones I had on Hedron (my XO). I'm not seeing anything glaring obvious that would be different. My only thoughts could be that maybe something on the flashed image was bad or similar to mess up how it's working.

It's hard for me to know for sure, but I'm totally befuddled now about what the issue might be or if this is an issue. I suppose, either way, this script should still work on the edge cases where this a problem, but I'm not entirely sure. I know @ColdSauce owned the XO before me, but I'm not sure if he would know much more than me either about why `yum` wasn't able to work with those repo files.

jwflory commented:
Since this script still fits into the existing repository configuration and the script itself is not completely off the mark, I rebased this pull request and did some cleanup (I also implemented more of @ryansb's original feedback). I'm going to merge this PR and we can always fix up the scripts later.

:clapper:
